// express-simple-errors/lib/errors.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var BadRequest = exports.BadRequest = function (_Error) {
  _inherits(BadRequest, _Error);

  function BadRequest(msg) {
    _classCallCheck(this, BadRequest);

    var _this = _possibleConstructorReturn(this, (BadRequest.__proto__ || Object.getPrototypeOf(BadRequest)).call(this, msg));

    _this.message = msg || 'Bad Request';
    _this.code = 400;
    Error.captureStackTrace(_this);
    return _this;
  }

  return BadRequest;
}(Error);

var Unauthorized = exports.Unauthorized = function (_Error2) {
  _inherits(Unauthorized, _Error2);

  function Unauthorized(msg) {
    _classCallCheck(this, Unauthorized);

    var _this2 = _possibleConstructorReturn(this, (Unauthorized.__proto__ || Object.getPrototypeOf(Unauthorized)).call(this, msg));

    _this2.message = msg || 'Unauthorized';
    _this2.code = 401;
    Error.captureStackTrace(_this2);
    return _this2;
  }

  return Unauthorized;
}(Error);

var Forbidden = exports.Forbidden = function (_Error3) {
  _inherits(Forbidden, _Error3);

  function Forbidden(msg) {
    _classCallCheck(this, Forbidden);

    var _this3 = _possibleConstructorReturn(this, (Forbidden.__proto__ || Object.getPrototypeOf(Forbidden)).call(this, msg));

    _this3.message = msg || 'Forbidden';
    _this3.code = 403;
    Error.captureStackTrace(_this3);
    return _this3;
  }

  return Forbidden;
}(Error);

var NotFound = exports.NotFound = function (_Error4) {
  _inherits(NotFound, _Error4);

  function NotFound(msg) {
    _classCallCheck(this, NotFound);

    var _this4 = _possibleConstructorReturn(this, (NotFound.__proto__ || Object.getPrototypeOf(NotFound)).call(this, msg));

    _this4.message = msg || 'Not Found';
    _this4.code = 404;
    Error.captureStackTrace(_this4);
    return _this4;
  }

  return NotFound;
}(Error);

var Conflict = exports.Conflict = function (_Error5) {
  _inherits(Conflict, _Error5);

  function Conflict(msg) {
    _classCallCheck(this, Conflict);

    var _this5 = _possibleConstructorReturn(this, (Conflict.__proto__ || Object.getPrototypeOf(Conflict)).call(this, msg));

    _this5.message = msg || 'Conflict';
    _this5.code = 409;
    Error.captureStackTrace(_this5);
    return _this5;
  }

  return Conflict;
}(Error);

var UnsupportedMediaType = exports.UnsupportedMediaType = function (_Error6) {
  _inherits(UnsupportedMediaType, _Error6);

  function UnsupportedMediaType(msg) {
    _classCallCheck(this, UnsupportedMediaType);

    var _this6 = _possibleConstructorReturn(this, (UnsupportedMediaType.__proto__ || Object.getPrototypeOf(UnsupportedMediaType)).call(this, msg));

    _this6.message = msg || 'Unsupported Media Type';
    _this6.code = 415;
    Error.captureStackTrace(_this6);
    return _this6;
  }

  return UnsupportedMediaType;
}(Error);
